gossip unit test failing intermittently

For example https://travis-ci.org/weaveworks/weave/builds/66049086:

````
--- FAIL: TestGossipTopology-2 (0.00s)
	util.go:134: Expected connections not found: from 02:00:00:02:00:00 to map[03:00:00:03:00:00:{}]
...
	github.com/weaveworks/weave/router.implTestGossipTopology(0xc2091082d0)
		/home/travis/gopath/src/github.com/weaveworks/weave/router/gossip_test.go:116 +0xc0b
````

Appears to be caused by topology updates being delivered on a different goroutine, which sometimes happens later than expected.  There is a 'FIXME' comment to the effect that `sendPendingGossip()` may not work.

`go test -race` reports many race conditions in this test, which may or may not be relevant.

--- a/router/gossip.go
+++ b/router/gossip.go
@@ -332,15 +332,17 @@ func (router *Router) sendPendingGossip() {
 	for _, channel := range router.GossipChannels {
 		channel.Lock()
 		for _, sender := range channel.senders {
-			ch := make(chan struct{})
-			sender.flushch <- ch
-			<-ch
+			sender.flush()
 		}
 		for _, sender := range channel.broadcasters {
-			ch := make(chan struct{})
-			sender.flushch <- ch
-			<-ch
+			sender.flush()
 		}
 		channel.Unlock()
 	}
+}
+
+func (sender *GossipSender) flush() {
+	ch := make(chan struct{})
+	sender.flushch <- ch
+	<-ch
 }

--- a/router/gossip.go
+++ b/router/gossip.go
@@ -63,7 +63,14 @@ func (sender *GossipSender) run() {
 			}
 			sender.send(pending)
 		case ch := <-sender.flushch:
-			sender.flush()
+			// ensure anything pending is sent, then close the channel we were given
+			select {
+			case pending := <-sender.cell:
+				if pending != nil {
+					sender.send(pending)
+				}
+			default:
+			}
 			close(ch)
 		}
 	}
@@ -339,16 +346,5 @@ func (router *Router) sendPendingGossip() {
 		ch := make(chan struct{})
 		sender.flushch <- ch
 		<-ch
-	}
-}
-
-func (sender *GossipSender) flush() {
-	for {
-		select {
-		case pending := <-sender.cell:
-			sender.send(pending)
-		default:
-			return
-		}
 	}
 }

--- a/router/gossip.go
+++ b/router/gossip.go
@@ -39,8 +39,9 @@ type Gossiper interface {
 // Accumulates GossipData that needs to be sent to one destination,
 // and sends it when possible.
 type GossipSender struct {
-	send func(GossipData)
-	cell chan GossipData
+	send    func(GossipData)
+	cell    chan GossipData
+	flushch chan chan struct{}
 }
 
 func NewGossipSender(send func(GossipData)) *GossipSender {
@@ -49,15 +50,22 @@ func NewGossipSender(send func(GossipData)) *GossipSender {
 
 func (sender *GossipSender) Start() {
 	sender.cell = make(chan GossipData, 1)
+	sender.flushch = make(chan chan struct{})
 	go sender.run()
 }
 
 func (sender *GossipSender) run() {
 	for {
-		if pending := <-sender.cell; pending == nil { // receive zero value when chan is closed
-			break
-		} else {
-			sender.send(pending)
+		select {
+		case pending := <-sender.cell:
+			if pending == nil { // receive zero value when chan is closed
+				return
+			} else {
+				sender.send(pending)
+			}
+		case ch := <-sender.flushch:
+			sender.flush()
+			close(ch)
 		}
 	}
 }
@@ -314,17 +322,24 @@ func (c *GossipChannel) log(args ...interface{}) {
 
 // for testing
 
-// FIXME this doesn't actually guarantee everything has been sent
-// since a GossipSender may be in the process of sending and there is
-// no easy way for us to know when that has completed.
 func (router *Router) sendPendingGossip() {
+	// copy the senders so we don't hold the lock for long
+	allSenders := make([]*GossipSender, 0)
 	for _, channel := range router.GossipChannels {
+		channel.Lock()
 		for _, sender := range channel.senders {
-			sender.flush()
+			allSenders = append(allSenders, sender)
 		}
 		for _, sender := range channel.broadcasters {
-			sender.flush()
+			allSenders = append(allSenders, sender)
 		}
+		channel.Unlock()
+	}
+
+	for _, sender := range allSenders {
+		ch := make(chan struct{})
+		sender.flushch <- ch
+		<-ch
 	}
 }
 

--- a/router/gossip.go
+++ b/router/gossip.go
@@ -329,22 +329,18 @@ func (c *GossipChannel) log(args ...interface{}) {
 // for testing
 
 func (router *Router) sendPendingGossip() {
-	// copy the senders so we don't hold the lock for long
-	var allSenders []*GossipSender
 	for _, channel := range router.GossipChannels {
 		channel.Lock()
 		for _, sender := range channel.senders {
-			allSenders = append(allSenders, sender)
+			ch := make(chan struct{})
+			sender.flushch <- ch
+			<-ch
 		}
 		for _, sender := range channel.broadcasters {
-			allSenders = append(allSenders, sender)
+			ch := make(chan struct{})
+			sender.flushch <- ch
+			<-ch
 		}
 		channel.Unlock()
-	}
-
-	for _, sender := range allSenders {
-		ch := make(chan struct{})
-		sender.flushch <- ch
-		<-ch
 	}
 }

--- a/router/gossip.go
+++ b/router/gossip.go
@@ -60,9 +60,8 @@ func (sender *GossipSender) run() {
 		case pending := <-sender.cell:
 			if pending == nil { // receive zero value when chan is closed
 				return
-			} else {
-				sender.send(pending)
 			}
+			sender.send(pending)
 		case ch := <-sender.flushch:
 			sender.flush()
 			close(ch)
@@ -324,7 +323,7 @@ func (c *GossipChannel) log(args ...interface{}) {
 
 func (router *Router) sendPendingGossip() {
 	// copy the senders so we don't hold the lock for long
-	allSenders := make([]*GossipSender, 0)
+	var allSenders []*GossipSender
 	for _, channel := range router.GossipChannels {
 		channel.Lock()
 		for _, sender := range channel.senders {

--- a/router/gossip_test.go
+++ b/router/gossip_test.go
@@ -30,7 +30,6 @@ func (conn *mockChannelConnection) SendProtocolMsg(protocolMsg ProtocolMsg) {
 	if err := conn.dest.handleGossip(protocolMsg.tag, protocolMsg.msg); err != nil {
 		panic(err)
 	}
-	conn.dest.sendPendingGossip()
 }
 
 func (router *Router) AddTestChannelConnection(r *Router) {
@@ -87,7 +86,9 @@ func (router *Router) tp(routers ...*Router) *Peer {
 
 // Check that the topology of router matches the peers and all of their connections
 func checkTopology(t *testing.T, router *Router, wantedPeers ...*Peer) {
+	router.Peers.RLock()
 	checkTopologyPeers(t, true, router.Peers.allPeers(), wantedPeers...)
+	router.Peers.RUnlock()
 }
 
 func implTestGossipTopology(t *testing.T) {

